## 대표적인 데이터 구조6: 해쉬 테이블 (Hash Table)

### 1. 해쉬 구조
* Hash Table: 키(Key)에 데이터(Value)를 저장하는 데이터 구조
  - Key를 통해 바로 데이터를 받아올 수 있으므로, 속도가 획기적으로 빨라짐
  - 파이썬 딕셔너리(Dictionary) 타입이 해쉬 테이블의 예: Key를 가지고 바로 데이터(Value)를 꺼냄
  - 보통 배열로 미리 Hash Table 사이즈만큼 생성 후에 사용 (공간과 탐색 시간을 맞바꾸는 기법)
  - <font color='#BF360C'>단, 파이썬에서는 해쉬를 별도 구현할 이유가 없음 - 딕셔너리 타입을 사용하면 됨</font>

### 2. 알아둘 용어
* 해쉬(Hash): 임의 값을 고정 길이로 변환하는 것
* 해쉬 테이블(Hash Table): 키 값의 연산에 의해 직접 접근이 가능한 데이터 구조
* 해싱 함수(Hashing Function): Key에 대해 산술 연산을 이용해 데이터 위치를 찾을 수 있는 함수
* 해쉬 값(Hash Value) 또는 해쉬 주소(Hash Address): Key를 해싱 함수로 연산해서, 해쉬 값을 알아내고, 이를 기반으로 해쉬 테이블에서 해당 Key에 대한 데이터 위치를 일관성있게 찾을 수 있음
* 슬롯(Slot): 한 개의 데이터를 저장할 수 있는 공간
* 저장할 데이터에 대해 Key를 추출할 수 있는 별도 함수도 존재할 수 있음
<img src="https://www.fun-coding.org/00_Images/hash.png" width=400 />

### 3. 간단한 해쉬 예

#### 3.1. hash table 만들기
* 참고: 파이썬 list comprehension - https://www.fun-coding.org/PL&OOP5-2.html

In [1]:
hash_table = list([i for i in range(10)])
hash_table

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

#### 3.2. 이번엔 초간단 해쉬 함수를 만들어봅니다.
- 다양한 해쉬 함수 고안 기법이 있으며, 가장 간단한 방식이 Division 법 (나누기를 통한 나머지 값을 사용하는 기법)

In [2]:
def hash_func(key):
    return key % 5

#### 3.3. 해쉬 테이블에 저장해보겠습니다.
- 데이터에 따라 필요시 key 생성 방법 정의가 필요함

In [3]:
data1 = 'Andy'
data2 = 'Dave'
data3 = 'Trump'
data4 = 'Anthor'
## ord(): 문자의 ASCII(아스키)코드 리턴
print (ord(data1[0]), ord(data2[0]), ord(data3[0]))
print (ord(data1[0]), hash_func(ord(data1[0])))
print (ord(data1[0]), ord(data4[0]))

65 68 84
65 0
65 65


- 3.3.2. 해쉬 테이블에 값 저장 예
  - data:value 와 같이 data 와 value를 넣으면, 해당 data에 대한 key를 찾아서, 해당 key에 대응하는 해쉬주소에 value를 저장하는 예

In [4]:
def storage_data(data, value):
    key = ord(data[0])
    hash_address = hash_func(key)
    hash_table[hash_address] = value

#### 3.4. 해쉬 테이블에서 특정 주소의 데이터를 가져오는 함수도 만들어봅니다.

In [5]:
storage_data('Andy', '01055553333')
storage_data('Dave', '01044443333')
storage_data('Trump', '01022223333')

#### 3.5. 실제 데이터를 저장하고, 읽어보겠습니다.

In [6]:
def get_data(data):
    key = ord(data[0])
    hash_address = hash_func(key)
    return hash_table[hash_address]

In [7]:
get_data('Andy')

'01055553333'

### 4. 자료 구조 해쉬 테이블의 장단점과 주요 용도
- 장점
  - 데이터 저장/읽기 속도가 빠르다. (검색 속도가 빠르다.)
  - 해쉬는 키에 대한 데이터가 있는지(중복) 확인이 쉬움
- 단점 
  - 일반적으로 저장공간이 좀더 많이 필요하다.
  - **여러 키에 해당하는 주소가 동일할 경우 충돌을 해결하기 위한 별도 자료구조가 필요함**
- 주요 용도
  - 검색이 많이 필요한 경우
  - 저장, 삭제, 읽기가 빈번한 경우
  - 캐쉬 구현시 (중복 확인이 쉽기 때문)

### 5. 프로그래밍 연습

<div class="alert alert-block alert-warning">
<strong><font color="blue" size="3em">연습1: 리스트 변수를 활용해서 해쉬 테이블 구현해보기</font></strong><br>
1. 해쉬 함수: key % 8<br>
2. 해쉬 키 생성: hash(data)
</div>

In [8]:
hash_table = list([0 for i in range(8)])

def get_key(data):
    return hash(data)

def hash_function(key):
    return key % 8

def save_data(data, value):
    hash_address = hash_function(get_key(data))
    hash_table[hash_address] = value

def read_data(data):
    hash_address = hash_function(get_key(data))
    return hash_table[hash_address]

In [9]:
print(hash_table)
save_data('Dave', '0102030200')
save_data('Andy', '01033232200')
read_data('Dave')

[0, 0, 0, 0, 0, 0, 0, 0]


'0102030200'

In [10]:
print(read_data('Dave'))
print(read_data('Andy'))

0102030200
01033232200


In [11]:
hash_table

['0102030200', 0, 0, 0, 0, '01033232200', 0, 0]

### 6. 충돌(Collision) 해결 알고리즘 (좋은 해쉬 함수 사용하기)
> 해쉬 테이블의 가장 큰 문제는 충돌(Collision)의 경우입니다.
> 이 문제를 충돌(Collision) 또는 해쉬 충돌(Hash Collision)이라고 부릅니다.

#### 6.1. Chaining 기법
- **개방 해슁 또는 Open Hashing 기법** 중 하나: 해쉬 테이블 저장공간 외의 공간을 활용하는 기법
- 충돌이 일어나면, 링크드 리스트라는 자료 구조를 사용해서, 링크드 리스트로 데이터를 추가로 뒤에 연결시켜서 저장하는 기법

<div class="alert alert-block alert-warning">
<strong><font color="blue" size="3em">연습2: 연습1의 해쉬 테이블 코드에 Chaining 기법으로 충돌해결 코드를 추가해보기</font></strong><br>
1. 해쉬 함수: key % 8<br>
2. 해쉬 키 생성: hash(data)
</div>

In [12]:
hash_table = list([0 for i in range(8)])

def get_key(data):
    return hash(data)

def hash_function(key):
    return key % 8

def save_data(data, value):
    index_key = get_key(data)
    hash_address = hash_function(index_key)
    # hash_table을 0으로 초기화했기 때문에, value가 들어있다면 0이 아닐 것이라고 가정
    # 즉, 0이 아니라면 해당 위치에 데이터가 들어가 있다는 것
    if hash_table[hash_address] != 0:
        # 파이썬에서는 리스트에 append로 데이터를 추가하면 linkedlist와 유사한 효과를 낸다.
        # 특정 주소에 데이터가 하나라도 들어가 있다면 그것이 리스트 타입으로 구현되어 있음.
        # 따라서, len(hash_table[hash_address])를 통해서 해당 주소에 데이터가 하나가 들어가 있는지 2개가 들어있는지
        # 들어간 데이터의 개수를 알아볼 수 있다.
        for index in range(len(hash_table[hash_address])):
            # Chaining에서 동일한 hash를 가질 경우 기존에 있던 값 뒤에 이어서
            # linkedlist 형태로 [key, value] 쌍을 저장해나간다.
            # 따라서, index는 몇 번째로 이어져 있는 [key, value] 쌍인지 확인하는 용도.
            # hash_table[hash_address][index][0] == index_key 인 경우는 기존에 있던 값을 갱신하는 것.
            if hash_table[hash_address][index][0] == index_key:
                hash_table[hash_address][index][1] = value
                # 의도한 작업을 마쳤다면 함수를 빠져나감
                return
        # hash_table[hash_address][index][0] != index_key 인 경우는 기존에 있던 값 뒤에 이어서 새로운 값을 추가하는 것.
        hash_table[hash_address].append([index_key, value])
    # hash_address 주소를 가지는 위치에 어떤 데이터도 저장되어 있지 않다면
    # [index_key, value] 형태를 가지는 리스트를 추가한다.
    else:
        hash_table[hash_address] = [[index_key, value]]
    
def read_data(data):
    index_key = get_key(data)
    hash_address = hash_function(index_key)

    # 해당 위치에 데이터가 있는 경우
    if hash_table[hash_address] != 0:
        for index in range(len(hash_table[hash_address])):
            if hash_table[hash_address][index][0] == index_key:
                return hash_table[hash_address][index][1]
        return None
    # 해당 위치에 데이터가 없다면 아무 것도 반환하지 않는다.
    else:
        return None


In [13]:
print (hash('Dave') % 8)
print (hash('Dd') % 8)
print (hash('Data') % 8)

0
0
3


In [16]:
save_data('Dd', '1201023010')
save_data('Dave', '3301023010')
read_data('Dd')

'1201023010'

In [17]:
hash_table

[[[8145393072902308328, '1201023010'], [-8375584257681269872, '3301023010']],
 0,
 0,
 0,
 0,
 0,
 0,
 0]

위의 결과를 보면 `[[8145393072902308328, '1201023010'], [-8375584257681269872, '3301023010']]`과 같이 데이터를 저장하고 있는 모습을 볼 수 있는데, `8145393072902308328`은 key이고 `'1201023010'`이 대응하는 value라고 보면 된다.

만약, 배열로 Chaining을 한다면 충돌이 얼마나 발생할지 예측할 수 없기 때문에 최대 충돌 횟수만큼 미리 공간을 잡아두는 것이 어렵다. 반면, linkedlist는 충돌이 일어날 때만 리스트 데이터를 생성하면 되기 때문에 가장 적절한 구현방법이 될 수 있다.

#### 6.2. Linear Probing 기법
- **폐쇄 해슁 또는 Close Hashing 기법** 중 하나: 해쉬 테이블 저장공간 안에서 충돌 문제를 해결하는 기법
- 충돌이 일어나면, 해당 hash address의 다음 address부터 맨 처음 나오는 빈공간에 저장하는 기법
  - 저장공간 활용도를 높이기 위한 기법

<div class="alert alert-block alert-warning">
<strong><font color="blue" size="3em">연습3: 연습1의 해쉬 테이블 코드에 Linear Probling 기법으로 충돌해결 코드를 추가해보기</font></strong><br>
1. 해쉬 함수: key % 8<br>
2. 해쉬 키 생성: hash(data)
</div>

In [18]:
hash_table = list([0 for i in range(8)])

def get_key(data):
    return hash(data)

def hash_function(key):
    return key % 8

def save_data(data, value):
    index_key = get_key(data)
    hash_address = hash_function(index_key)

    # 이미 데이터가 저장되어 있는 슬롯인 경우
    if hash_table[hash_address] != 0:
        # 이미 있는 데이터가 저장하려는 데이터와 같은지, 다른지 확인한다.
        # 현재 주소를 기준으로 다음 주소에 빈공간이 있는지 확인해야 한다. 
        for index in range(hash_address, len(hash_table)):
            # 현재 주소에 담긴 데이터가 0인 경우 = 빈 공간
            if hash_table[index] == 0:
                # 해당 빈 공간에 리스트 형태로 key와 value를 저장한다.
                hash_table[index] = [index_key, value]
                # 저장하는 작업이 끝났기 때문에 함수를 탈출
                return
            # 동일한 인덱스 키를 가진다면
            # 새로 공간을 마련할 것이 아니라, 기존의 값을 업데이트해야 하는 것이다.
            elif hash_table[index][0] == index_key:
                hash_table[index][1] = value
                return
    # 해당 슬롯에는 데이터가 저장된 적이 없음.
    # 충돌이 아닌 상황이기 때문에 해당 위치에 저장하면 된다.
    else:
        hash_table[hash_address] = [index_key, value]

def read_data(data):
    index_key = get_key(data)
    hash_address = hash_function(index_key)
    
    if hash_table[hash_address] != 0:
        for index in range(hash_address, len(hash_table)):
            # 동일한 해시를 가져서 충돌이 발생한 경우 Linear Probing 기법에서는 다음 빈공간을 찾아서 저장을 한다.
            # 그런데, hash_table[index] == 0인 경우는 for문을 돌면서 계속해서 원하는 값을 찾지 못하다가
            # 빈 공간을 찾게 된 경우이기 때문에
            # 이러한 경우는 읽으려는 데이터가 한 번도 테이블에 저장된 적이 없다는 것을 의미한다.
            # 따라서, 찾으려는 데이터가 없는 상태를 None을 리턴하는 것으로 처리한다.
            if hash_table[index] == 0:
                return None
            # 내가 찾으려는 데이터를 찾았다면 해당 위치의 value를 리턴한다.
            elif hash_table[index][0] == index_key:
                return hash_table[index][1]

    # 찾으려는 데이터가 없는 상태이기 때문에 None을 리턴하는 것으로 처리한다.
    else:
        return None


In [25]:
print (hash('da') % 8)
print (hash('dc') % 8)

5
5


In [26]:
save_data('dc', '01200123123')
save_data('da', '3333333333')
read_data('dc')

'01200123123'

#### 6.3. 빈번한 충돌을 개선하는 기법
- 해쉬 함수을 재정의 및 해쉬 테이블 저장공간을 확대
- 예:

```python
hash_table = list([None for i in range(16)])

def hash_function(key):
    return key % 16
```

### 참고: 해쉬 함수와 키 생성 함수
- 파이썬의 hash() 함수는 실행할 때마다, 값이 달라질 수 있음
- 유명한 해쉬 함수들이 있음: SHA(Secure Hash Algorithm, 안전한 해시 알고리즘)
  - 어떤 데이터도 유일한 고정된 크기의 고정값을 리턴해주므로, 해쉬 함수로 유용하게 활용 가능

#### SHA-1

In [27]:
import hashlib

# 'test'라는 문자열을 인코딩
# 즉, 바이트로 바꾸어주는 작업을 말함.
# .encode()를 사용하지 않고, 직접 hash_object.update(b'test')와 같이 사용할 수도 있음.
data = 'test'.encode()
hash_object = hashlib.sha1()
hash_object.update(data)
hex_dig = hash_object.hexdigest()
print (hex_dig)

a94a8fe5ccb19ba61c4c0873d391e987982fbbd3


#### SHA-256

In [28]:
import hashlib

data = 'test'.encode()
hash_object = hashlib.sha256()
hash_object.update(data)
hex_dig = hash_object.hexdigest()
print (hex_dig)

9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08


<div class="alert alert-block alert-warning">
<strong><font color="blue" size="3em">연습4: 연습2의 Chaining 기법을 적용한 해쉬 테이블 코드에 키 생성 함수를 sha256 해쉬 알고리즘을 사용하도록 변경해보기</font></strong><br>
1. 해쉬 함수: key % 8<br>
2. 해쉬 키 생성: hash(data)
</div>

In [29]:
import hashlib

hash_table = list([0 for i in range(8)])

def get_key(data):
        hash_object = hashlib.sha256()
        hash_object.update(data.encode())
        hex_dig = hash_object.hexdigest()
        # hex_dig은 문자열 형태로 되어 있는데
        # 실제로는 16진수 형태로 표현된 것이고,
        # 해시함수를 적용하기 위해서 int형태(10진수)로 변환한다.
        return int(hex_dig, 16)

def hash_function(key):
    return key % 8

def save_data(data, value):
    index_key = get_key(data)
    hash_address = hash_function(index_key)
    if hash_table[hash_address] != 0:
        for index in range(hash_address, len(hash_table)):
            if hash_table[index] == 0:
                hash_table[index] = [index_key, value]
                return
            elif hash_table[index][0] == index_key:
                hash_table[index][1] = value
                return
    else:
        hash_table[hash_address] = [index_key, value]

def read_data(data):
    index_key = get_key(data)
    hash_address = hash_function(index_key)
    
    if hash_table[hash_address] != 0:
        for index in range(hash_address, len(hash_table)):
            if hash_table[index] == 0:
                return None
            elif hash_table[index][0] == index_key:
                return hash_table[index][1]
    else:
        return None


In [31]:
print (get_key('da') % 8)
print (get_key('dh') % 8)

2
2


In [32]:
save_data('da', '01200123123')
save_data('dh', '3333333333')
read_data('dh')

'3333333333'

### 7. 시간 복잡도
- 일반적인 경우(Collision이 없는 경우)는 O(1)
- 최악의 경우(Collision이 모두 발생하는 경우)는 O(n)

> 해쉬 테이블의 경우, 일반적인 경우를 기대하고 만들기 때문에, 시간 복잡도는 O(1) 이라고 말할 수 있음

### 검색에서 해쉬 테이블의 사용 예
- 16개의 배열에 데이터를 저장하고, 검색할 때 O(n)
- 16개의 데이터 저장공간을 가진 위의 해쉬 테이블에 데이터를 저장하고, 검색할 때 O(1)